### DISCUSSION TITLE
Using aggregate result in the observer query?

### DISCUSSION THREAD

**Discussion by Falkovic1993** [robfallows/tunguska-reactive-aggregate#59] [General]
Hey Rob, 

Thanks for the package, works really well 👍 

I have a question regarding performance optimizing, when using the reactive aggregate to lookup in another collection would it be possible to use the result of the aggregate in the observer if you have multiple observers? See example here; 

`
 ReactiveAggregate(
      this,
      Bookings,
      [
        {
          $match: {
            restaurantId,
            dateTime: { $gte: fromDateTime, $lt: toDateTime },
          },
        },
        {
          $lookup: {
            from: 'Customers',
            localField: 'customerId',
            foreignField: '_id',
            as: 'customer',
          },
        },
        { $unwind: '$customer' },
        { $project: neededBookingFields },
      ],
      {
        noAutomaticObserver: true,
        debounceCount: 100,
        debounceDelay: 100,
        observers: [Bookings.find(query), Customers.find({ restaurantId })]
      }
    );
    `
   
Since we want to observe both our bookings as well as the customers, it would make sense to only observe those customers who are matching the bookings. In our example, we save our customerId on the bookings. So let's say we're looking at a day with 10 bookings, the maximum customers we should observe would be the same or less, but instead, we observing all customers, which is not optimal in the long run :-) 

Hope it makes sense, else let me know if I should elaborate a bit more. 

Cheers

**Comment by robfallows** [comment]
Hi @Falkovic1993 - glad you're finding the package useful 🙂 

I'm not sure I understand well enough to comment. However:

> In our example, we save our customerId on the bookings.

Isn't it just a case of ensuring that the `Bookings.find(query)` passed to the `observers` basically matches the pipeline's initial match (`restaurantId`, `datetime`) anded with the `customerId`? In other words, only observe the subset of docs fulfilling _all_ the criteria.

**Reply by Falkovic1993** [reply]
Thanks for the quick answer @robfallows 

In this case, it should be in connection to the `Customers.find({restaurantId })` since we want to restrict the number of customers we observe, so my idea was kinda if there was a way to get the results from the aggregate to use so the customers observe would look something like this `Customers.find({resturantId, "_id" : { $in : [ customerIdsFromBookingsAggregateResult ] } `

First, when we know which bookings from the $match based on the date range, we can get an idea of which customers we want to observe.

**Reply by robfallows** [reply]
🤔 In a single reactive aggregate publication? I'm thinking not possible because publications are not themselves reactive. So once the initial observers are in place they can't be changed (any changes won't be acted upon) unless the publication is stopped and restarted. That means you have to know exactly what you're observing at the time of setting up the publication.

Happy to be proven wrong if anyone else has a solution.

**Reply by Falkovic1993** [reply]
Yeah makes a lot of sense, I don't think what I'm trying to do is possible - So I'll probably change the data structure a bit to fix the issue in the long run. 

Again, thanks a lot for the response and the package!

Cheers
